### PR TITLE
Add transformer to LM example

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

1. Add transformer to PTB LM example
2. Add linear sharing weights in Transformer

## How was this patch tested?

unit tests
example tests

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBWordLM.scala [hunk 1]
```diff
@@ -66,15 +66,20 @@ object PTBWordLM {
...
+      } else if (param.withTransformerModel) {
+        PTBModel.transformer(
+          inputSize = param.vocabSize,
+          hiddenSize = param.hiddenSize,
+          outputSize = param.vocabSize,
+          numLayers = param.numLayers,
+          keepProb = param.keepProb)
       } else {
-        val curModel = PTBModel(
+        PTBModel(
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBWordLM.scala]: PTBModel.lstm ? we need to keep consistent in terms of  different model naming

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBModel.scala [hunk 2]
```diff
@@ -21,6 +21,23 @@ import com.intel.analytics.bigdl.nn.Graph._
...
+    val transformer = Transformer[Float](vocabSize = inputSize,
+      hiddenSize = hiddenSize, numHeads = 4, filterSize = hiddenSize*4,
+      numHiddenlayers = numLayers, embeddingDropout = 1- keepProb,
+      attentionDropout = 0.1f, ffnDropout = 0.1f).inputs(input)
+    val linear = Linear[Float](hiddenSize, outputSize)
+    val output = TimeDistributed[Float](linear).inputs(transformer)
+    Graph(input, output)
+  }
+
   def apply(
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBModel.scala]: the above one is based on transformer, do we need to change this to lstm ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TableOperation.scala [hunk 3]
```diff
@@ -55,7 +55,10 @@ class TableOperation[T: ClassTag](
     val inputSmall = input[Tensor[T]](smallPos)
     val inputLarge = input[Tensor[T]](3 - smallPos)
 
-    if (expandLayer == null) expandLayer = ExpandSize(inputLarge.size())
+    val largeSize = inputLarge.size()
+    // batchSize may be not same for non-training case
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TableOperation.scala]: what do you mean by `non-training`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBWordLM.scala (hunk 1 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBWordLM.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBModel.scala (hunk 2 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/example/languagemodel/PTBModel.scala]: ok, change to lstm

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TableOperation.scala (hunk 3 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TableOperation.scala]: ok, maybe Inference is more clear. The size of last batch may not be same with previous ones for model Inference

**Comment by zhangxiaoli73** [comment]
jenkins pass: http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/648/

**Pull request merged**